### PR TITLE
SLING-8595 update to bundle-parent 40

### PR DESCRIPTION
**Pull request by joerghoh**
the majority of the PR is the migration from SCR annotations to OSGI annotations. While everything seems to work quite well, I am not really confident, because the generated DS and metatype files seem to differ a lot.

### CODE REVIEW

File: src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java [hunk 1]
```diff
@@ -22,13 +22,15 @@
...
 import org.apache.sling.distribution.queue.DistributionQueueItem;
+import org.junit.Ignore;
 import org.junit.Test;
 
 import static org.junit.Assert.*;
 
 public class QueueItemMapperTest {
 
     @Test
+    @Ignore
```
Comment (tmaret) [inline src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: Is the `Ignore` overlooked or required ?

File: src/main/java/org/apache/sling/distribution/agent/impl/ForwardDistributionAgentFactory.java [hunk 2]
```diff
@@ -69,125 +59,136 @@
...
  */
-@Component(metatype = true,
-        label = "Apache Sling Distribution Agent - Forward Agents Factory",
-        description = "OSGi configuration factory for forward agents",
-        configurationFactory = true,
-        specVersion = "1.1",
-        policy = ConfigurationPolicy.REQUIRE
+@Component(
+        configurationPolicy = ConfigurationPolicy.REQUIRE,
+        property= {"webconsole.configurationFactory.nameHint=Agent name: {name}"}
```
Comment (tmaret) [inline src/main/java/org/apache/sling/distribution/agent/impl/ForwardDistributionAgentFactory.java:79]: An alternative to keep all props in the Config, is to define the webconsole.configurationFactory.nameHint like in 

https://github.com/apache/sling-org-apache-sling-distribution-journal/blob/02846937fa0e07d183f50484bc68e028337c3fe5/src/main/java/org/apache/sling/distribution/journal/impl/publisher/PublisherConfiguration.java#L28-L29

File: pom.xml [hunk 3]
```diff
@@ -59,43 +58,47 @@
...
-                            org.apache.sling.event.jobs.consumer;version="[1.1,2)",
-                            org.apache.sling.jcr.api;version="2.2.0",
-                            *
-                        </Import-Package>
-                    </instructions>
-                </configuration>
+                <groupId>biz.aQute.bnd</groupId>
+                <artifactId>bnd-maven-plugin</artifactId>
             </plugin>
+<!--             <plugin> -->
```
Comment (rombert) [inline pom.xml]: You should remove the commented out code.

File: pom.xml [hunk 4]
```diff
@@ -159,21 +162,39 @@
...
+        <dependency>
+        	<groupId>org.osgi</groupId>
+        	<artifactId>org.osgi.service.component.annotations</artifactId>
+        	<scope>provided</scope>
+      	</dependency>
+      	<dependency>
+        	<groupId>org.osgi</groupId>
+        	<artifactId>org.osgi.service.metatype.annotations</artifactId>
+        	<scope>provided</scope>
+      	</dependency>
```
Comment (rombert) [inline pom.xml]: Nitpick: indentation is off here.

File: src/main/java/org/apache/sling/distribution/impl/DefaultDistributor.java [hunk 5]
```diff
@@ -31,22 +28,22 @@
...
+import org.osgi.service.component.annotations.Reference;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Default implementation of Distributor interface that dispatches the request to available agents.
  */
-@Component
-@Service(Distributor.class)
+@Component(service=Distributor.class)
```
Comment (rombert) [inline src/main/java/org/apache/sling/distribution/impl/DefaultDistributor.java]: IIRC you don't need `service=...` here as it's inferred.

File: src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java (hunk 1 above)
Comment (rombert) [inline src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: This should not be ignored.

**Review by rombert — CHANGES_REQUESTED** [review]
Thanks a lot @joerghoh ! I only reviewed the build changes, so I think someone closed to the project should review. In the meantime, can you please use https://issues.apache.org/jira/browse/SLING-8595 for the commit titles?

File: src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java (hunk 1 above)
Comment (joerghoh) [reply src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: The current master branch wasn't running properly when I compiled it. But sure, I will investigate.

Comment (rombert) [reply src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: Ack. Perhaps @cschneider / @tmaret have a bette idea.

Comment (joerghoh) [reply src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: The problem here is a [locale dependent parsing of numbers in QueueMapperImpl](https://github.com/apache/sling-org-apache-sling-distribution-core/blob/master/src/main/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapper.java#L55). We can defintely hardocde it to English locale (and have the "." as decimal separator) to fix it.

File: src/main/java/org/apache/sling/distribution/impl/DefaultDistributor.java (hunk 5 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/distribution/impl/DefaultDistributor.java]: fixed

File: pom.xml (hunk 4 above)
Comment (joerghoh) [reply pom.xml]: fixed

File: pom.xml (hunk 3 above)
Comment (joerghoh) [reply pom.xml]: fixed

File: src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java (hunk 1 above)
Comment (joerghoh) [reply src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: I hardcoded the locale now to English. As we are only parsing numbers, I don't consider the impact too big. Nevertheless it's an issue in the existing code and this should be considered a changed behavior.

File: src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionComponentProvider.java [hunk 6]
```diff
@@ -40,24 +33,17 @@
...
-        @Reference(name = "distributionQueueDistributionStrategy", referenceInterface = DistributionQueueDispatchingStrategy.class, cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE, policy = ReferencePolicy.DYNAMIC),
-        @Reference(name = "distributionTransportSecretProvider", referenceInterface = DistributionTransportSecretProvider.class, cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE, policy = ReferencePolicy.DYNAMIC),
-        @Reference(name = "distributionTrigger", referenceInterface = DistributionTrigger.class, cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE, policy = ReferencePolicy.DYNAMIC),
-        @Reference(name = "distributionRequestAuthorizationStrategy", referenceInterface = DistributionRequestAuthorizationStrategy.class, cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE, policy = ReferencePolicy.DYNAMIC),
-        @Reference(name = "distributionPackageBuilder", referenceInterface = DistributionPackageBuilder.class, cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+@Component ( property = {
```
Comment (tmaret) [inline src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionComponentProvider.java:44]: This must expose the `DistributionComponentProvider` service

File: src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionConfigurationManager.java [hunk 7]
```diff
@@ -42,24 +43,27 @@
...
-    @Property(label = "Resource Config Properties", description = "The resource config properties", value = {"enabled"})
-    public static final String CONFIG_PROPERTIES = "resource.config.properties";
-
-    @Property(label = "Resource Config Defaults", description = "The default values for resource config properties", value = {"serializationType=distribution"})
-    public static final String CONFIG_DEFAULTS = "resource.config.defaults";
+    
+    @ObjectClassDefinition()
+    public @interface Config {
+        @AttributeDefinition(name="Resource Config enabled",description = "If storing config in resource tree is enabled.")
+        boolean enabled() default false;
```
Comment (tmaret) [inline src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionConfigurationManager.java]: The prop must be named `resource_config_enabled` to be compatible

**Comment by joerghoh** [comment]
Let's try to get this merged, and then we will do an update to the most recent version

**Comment by rombert** [comment]
Sounds good to me. I think @cschneider needs to find the time to review this, not sure if someone else can help.

**Comment by joerghoh** [comment]
I raised https://issues.apache.org/jira/browse/SLING-11137 to handle the sonar issue.

**Review by tmaret — CHANGES_REQUESTED** [review]

**Comment by tmaret** [comment]
I forgot to submit my review earlier, sorry.

File: src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionComponentProvider.java (hunk 6 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionComponentProvider.java:44]: By default a ```Component``` registers to all interfaces which the class is implementing.

File: src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java (hunk 1 above)
Comment (joerghoh) [reply src/test/java/org/apache/sling/distribution/queue/impl/simple/QueueItemMapperTest.java]: fixed

File: src/main/java/org/apache/sling/distribution/agent/impl/ForwardDistributionAgentFactory.java (hunk 2 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/distribution/agent/impl/ForwardDistributionAgentFactory.java:79]: extracted into a dedicated file

File: src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionConfigurationManager.java (hunk 7 above)
Comment (joerghoh) [reply src/main/java/org/apache/sling/distribution/component/impl/DefaultDistributionConfigurationManager.java]: done

**Review by cschneider — CHANGES_REQUESTED** [review]
I have checked the names of all config properties. Sometime webconsole namehint seems to be messing but that is not a big issue I think.
I have also fixed the package imports. So now the new bundle can be install in the cloud too.
During testing I found though that many DS components do not come up with a xml parsing error. I am digging to understand what it means.

**Comment by cschneider** [comment]
Found the issue. We need to set property not properties.

**Review by cschneider — APPROVED** [review]

**Pull request merged**